### PR TITLE
Ownership utility

### PR DESCRIPTION
**Pull request by howard-stearns**
Provides a specific working case of decentralized "ownership" agreement, for discussion.
The only immediate goal is to be able to handle exclusivity among cooperating scripts in both repeating code and event handlers.
It is an open question as to whether we should provide this, and if so, to do so as is, or with some improvement or increased capability.

The ownership code is packaged into libraries/owner.js and is documented in that file. It is entirely in Javacript, without C++ changes.
* tests/performance/tribbles.js is the original example, in which repeating code in an entityScript is executed by only one participant.
* entityScripts/keepAway.js shows two use cases in one script, and is documented in the code.

As is, the code does not guarantee that there will be no overlapping edits, and the responsiveness in obtaining ownership might not be good enough for event handlers. Plausible improvments, either before or after merging are:
* Improve reponsiveness/robustness by exposing the from-server timestamps to Javascript.
* Improve responsiveness/robustness using authoritive winner-picking at Entity Server. (Does not require the preceding.)
* Enforce property editing ownership at server. (Requires preceding authoritive winner-picking, as well as changes to owner.js API.)
* Permissions, or more precisely, a capability system in which keys can be obtained through any permission tracking system. (Requires preceding enforcement.)

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 8357b0ce359a32c834dd8ed4150c2737ae838c79 available at: https://gist.github.com/af1013ba92e72063d77c<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR6903.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR6903.dmg

**Comment by howard-stearns** [comment]
https://app.asana.com/0/32622044445063/82421953908844
https://app.asana.com/0/32622044445063/82733589935858

**Comment by howard-stearns** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 8357b0ce359a32c834dd8ed4150c2737ae838c79 available at: https://gist.github.com/af1013ba92e72063d77c<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR6903.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/HighFidelity-Alpha-PR6903.dmg

**Comment by howard-stearns** [comment]
I think I can make this reliable and not dependent on timeouts or conflicting userData by using Paxos and our message server. API doesn't change, though.

File: examples/libraries/owner.js [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+
+
+// Answers a an object with the exported functions, and accepts an object that can either be the key string or an object to be
+// that the exported functions will be added to (and from which non-default parameters can be specified).
+// (The 'ownerModule =' is required due to the way Script.include works.)
+ownerModule = function ownerModule(exportsObjectOrKey) {
+    var exports = ('string' === typeof exportsObjectOrKey) ? {} : exportsObjectOrKey,
+        // Script can call ownerModule with different keys.
+        key = (exports === exportsObjectOrKey) ? exports.keyString : exportsObjectOrKey,
+        OWNERSHIP_DATA_KEY = exports.ownershipDataKey || 'io.highFidelity.owner', // Must be unique among user data.
```
Comment (howard-stearns) [inline examples/libraries/owner.js]: Alternatively, we could close over the entityId as well as the key. However, I wanted to keep renewingEdit() looking just like Entities.editEntities, so the entityId is passed in each call.

**Comment by howard-stearns** [comment]
Closing this, to do a simpler and more correct version just so the example isn't screwed up at launch.
https://app.asana.com/0/32622044445063/85114103526644

**Pull request closed without merging**